File: leetcode/top_k_frequent_words.go
// https://leetcode.com/problems/top-k-frequent-words/solution/
// 692. Top K Frequent Words
package main

import (
	"fmt"
	"sort"
)

type Count struct {
	word string
	freq int
}

func topKFrequent(words []string, k int) []string {
	freqMap := make(map[string]int)
	topKWords := make([]string, 0)
	for _, word := range words {
		freqMap[word] += +1
	}
	wordCount := make([]Count, 0)
	for key, val := range freqMap {
		wordCount = append(wordCount, Count{
			word: key,
			freq: val,
		})
	}
	sort.Slice(wordCount, func(i, j int) bool {
		return wordCount[i].freq > wordCount[j].freq || (wordCount[i].freq == wordCount[j].freq && wordCount[i].word < wordCount[j].word)
	})
	for i := 0; i < k; i++ {
		topKWords = append(topKWords, wordCount[i].word)
	}
	return topKWords
}

func main() {
	a := []string{"sa", "ka", "ssaa", "dsa", "dsa", "ka", "a", "b", "b"}
	k := 2
	res := topKFrequent(a, k)
	fmt.Println(res)
}

File: leetcode/top_k_frequent_words.py
from collections import Counter
from typing import List


class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        freqMap = Counter(words)
        arr = [(freqMap[x], x) for x in freqMap]
        # people.sort(key=lambda x: (-x[0], x[1]))
        arr.sort(key=lambda x: (-x[0], x[1]))

        return [arr[i][1] for i in range(k)]


if __name__ == '__main__':
    words = ["i", "love", "leetcode", "i", "love", "coding"]
    k = 2
    res = Solution().topKFrequent(words, k)
    print(res)
